### PR TITLE
[Tesla] Support for not persisting Tesla credentials

### PR DESCRIPTION
**Pull request by ngrodum**
For those not wanting to store the Tesla username+pw (myself being the first customer), I have added support for using and persisting the OAUTH access/refresh token pair. I have tried to minimize the impact of the change, and that led the code to use the refresh token on every startup (not strictly needed) and on 401s.

I have updated the readme, read more in the "Alternative secure configuration" section.

I have just started with openHAB a couple of days ago, and this is my first PR here.

@kgoderis

### CODE REVIEW

**Comment by kgoderis** [comment]
@ngrodum Nicolai, this extension is ok. However, if you do not mind I would like to integrate this a bit later. Lately I am struggling with the connectivity to the Tesla back-end servers. Sometimes thing change at their side with connectivity getting lost, events not being sent and so forth. It would be of great help if you as well could TRACE the Tesla binding and share your findings in order to nail this.

**Comment by ThomDietrich** [comment]
@kgoderis why is the one blocking the other?

**Comment by ngrodum** [comment]
Are the connectivity problems throttling from Tesla's side?
If so, I guess not creating excessive refreshtokens could be seen as a step for us to conform as a less abusive client...
If you use the username/password you add a new refresh token to your account that has a life time of like 3 months. (I don't really know how often you run the auth code tho).

I'll see if I can look at some traces (I will run the code in this PR - I don't want to store the password) - I'd prefer it if you could merge this and I could checkout the code you're working on.

**Comment by kgoderis** [comment]
I get a lot of time outs, service unavailable but not related to the auth.

**Comment by kgoderis** [comment]
Btw you should actually run it against the other tesla PR as that one fixes an error. But it is not yet merged as I am doing further tests.

**Comment by kgoderis** [comment]
> @kgoderis why is the one blocking the other?

The binding as it in the repo has a nasty bug, and causing too many reconnections to the Tesla backend. And I asked Kai to put my PR on hold as I am testing. 

This PR can be merged as far as I am concerned (Code is LGTM), only be aware that the binding is not ok right now

File: addons/binding/org.openhab.binding.tesla/OSGI-INF/TeslaCommands.xml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    are made available under the terms of the Eclipse Public License v1.0
+    which accompanies this distribution, and is available at
+    http://www.eclipse.org/legal/epl-v10.html
+
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="org.openhab.binding.tesla.command.TeslaCommandExtension">
+
+   <scr:implementation class="org.openhab.binding.tesla.command.TeslaCommandExtension"/>
+
+   <scr:reference bind="setStorageService"
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tesla/OSGI-INF/TeslaCommands.xml]: We usually also always define an unbind method `unsetStorageService`, in which we unset the internal field. Could you add that as well?

File: addons/binding/org.openhab.binding.tesla/OSGI-INF/TeslaHandlerFactory.xml [hunk 2]
```diff
@@ -11,7 +11,13 @@
 -->
 <scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="org.openhab.binding.tesla.internal.TeslaHandlerFactory">
 
-   <implementation class="org.openhab.binding.tesla.internal.TeslaHandlerFactory"/>
+   <scr:implementation class="org.openhab.binding.tesla.internal.TeslaHandlerFactory"/>
+
+   <scr:reference bind="setStorageService"
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tesla/OSGI-INF/TeslaHandlerFactory.xml]: same here

File: addons/binding/org.openhab.binding.tesla/README.md [hunk 3]
```diff
@@ -13,6 +13,20 @@ Currently, the Tesla Model S is supported by this binding.
 
 The binding uses a the owner's email address and password in order to access the Tesla Owners Remote API. This is the same email address and password as used in the Tesla smartphone app, that allows to remotely control the Tesla Model S
 
+## Alternative secure configuration
+
+After a successful connection (Thing is online), you can clear the username and password field from the configuration, leaving only the VIN. Instead of using the username and password, from now on, only the access token will be stored and used (access token + refresh token). This is a better approach for many reasons, one being that you cannot start the car without the password.
+If you really don't like the thought of your username and password being stored temporarily, there is also a third option: Leave username/password blank, go into the openHAB console (ssh -p 8101 openhab@localhost), and run the following command:
+```
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tesla/README.md]: please add an empty line before here (Jekyll requires it)

File: addons/binding/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/command/TeslaCommandExtension.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ * <p>
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.tesla/src/main/java/org/openhab/binding/tesla/command/TeslaCommandExtension.java]: Did you generated those headers through `mvn license:format`? Afaik, we do not have that `<p>` in there. Could you remove it here (and in the other modified headers)?

File: addons/binding/org.openhab.binding.tesla/OSGI-INF/TeslaCommands.xml (hunk 1 above)
Comment (ngrodum) [inline addons/binding/org.openhab.binding.tesla/OSGI-INF/TeslaCommands.xml]: coming right up :)

**Pull request merged**